rfcs/react-components/convergence/theme-shared-colors.md

```diff
@@ -0,0 +1,114 @@
+| 1200 tokens, alias references |                   70 |
+| 600 tokens                    |                   20 |
+| 200 tokens                    |                    7 |
+
+**The numbers confirm the performance decreases linearly depending on the number of CSS variables (and therefore theme tokens).**
+
+### New Shared Color Alias tokens request
+
+_The numbers might be inaccurate as the feature is not fully spec-ed yet!_
+
+In order to support new scenarios (colorful calendar view), there is a new request to add 3 new shared color alias tokens. Adding them would result in adding 3 alias tokens for each of 49 shared colors = adding 147 new shared color alias tokens which would affect application performance.
+
+## Detailed Design or Proposal
+
+The theme currently contains 49 global shared colors.
+
+Current alias structure forces each ramp to be uniform (the same set of alias tokens for each shared color), resulting in 49 alias tokens each time we need a new line item (new alias token). Today we have 9 alias slots = total alias count is 441 (49x9).
+
+Components have varying degrees of color needs.
+
+Instead of 1 uniform Shared alias table (today's experience), separate out into 3 tables:
+
+1. Semantic States
+   - There are 5 unique global colors needed to support this set (Danger, Success, Severe, Warning, Out of office)
+   - Assume 9 alias tokens (Could potentially remove 2 from this table as they were avatar specific)
+   - 45 alias tokens in total (5 x 9)
+2. Persona
+   - There are 30 unique global colors needed to support this set
+   - Only needs 3 alias tokens
+   - 90 tokens in total (30 x 3)
+3. Category (calendar categories) (not fully defined yet)
+   - 10 unique global colors to support this set
+   - Assuming 9 alias slots (subject to increase)
+   - 90 tokens in total (10 x 9)
+
+Dividing up by 3 tables would reduce token count from 441 to ~225 (Table 1+Table 2+Table 3) while adding support for Calendar scenario.
+
+### Pros and Cons
+
+- 👍 Reducing shared color alias tokens from 441 to ~225 instead of increasing them to ~600 (to add tokens needed for Calendar)
+- 👍 More clarity around which specific color to use for semantic states instead of relying on design guidance
+- 👍 Allows a bit more wiggle room for the semantic alias table to grow w/o adding too much bloat
+- 👎 **Breaking change after the final RC**
+- 👎 Three separate color sets instead of a single one
+
+## Discarded Solutions
+
+### Create `theme_v2` to maintain backwards compatibility
+
+Instead of breaking the current theme we can just deprecate it and add `_v2` version.
+
+That might cause partner confusion - releasing the first final version with deprecated things which, when used, would result in degraded performance. The breaking change should not affect partners much - the colors which we are removing are not supposed to be used anywhere. If those are, there will be a build time (TS) error to catch the problem quickly.
```

Comment:
if we wanna get this in without Breaking change in RC/stable IMO it's a feasible tradeoff. 

- "v1" theme will be marked with `@deprecated` , users having proper set of lint rules will be notified / vscode will visually notify as well by striking through the identifier

Question is if we can provide tooling to automigrate from v1 to v2. if yes we can go with the BC + tooling and provide documentation/guidance to partners how to execute migration.

Reply:
As long as we just remove tokens without renaming those which remain having v1 and v2 should be relatively easy - the library itself would use v2 tokens which would work in either v2 or v1.

If we decide to rename existing tokens, it would make the possibility of implementing it as a backwards compatible change much more difficult.
You have `redBackground1` in legacyTheme, `dangerBackground1` in the new theme - what are we going to use in our components?

We will need to have a token which will use new CSS variable and fallback to legacy CSS variable (`var(--dangerBackground1, --redBackground1)`). Or use token names in source code which do not map 1:1 to CSS variables in DOM.

For that reason, if design can give us the final tokens now, I would vote for breaking it + nice opportunity to setup our codemods infrastructure.

Reply:
+1 on the auto-migration tooling, perhaps even if we do it in a non-breaking way. I would like us to get good at authoring codemods for changes, it will help accelerate upgrade and adoption of new features.

Reply:
At the end we decided not to implement any migration tooling.
We decided to only remove tokens which are not supposed to be used in existing code. There is no way how to migrate the potential usages.